syntax = "proto3";

package urgences;

// --- SERVICE ---
service UrgenceService {
  // 1. Signaler une urgence (Unary)
  rpc SignalerUrgence (UrgenceRequest) returns (UrgenceResponse);

  // 2. Suivre une intervention en temps réel (Server Streaming)
  rpc SuivreIntervention (SuiviRequest) returns (stream SuiviUpdate);

  // 3. Récupérer l'historique (Unary)
  rpc HistoriqueUrgences (HistoriqueRequest) returns (HistoriqueResponse);
}

// --- ENUMS (Typage Fort) ---
enum TypeUrgence {
  INCONNU = 0;
  ACCIDENT_ROUTE = 1;
  INCENDIE = 2;
  MALAISE = 3;
  INONDATION = 4;
  FUITE_GAZ = 5;
}

enum StatutIntervention {
  EN_ATTENTE = 0;
  EN_COURS = 1;
  TERMINE = 2;
}

// --- MESSAGES ---
message Coordonnees {
  double latitude = 1;
  double longitude = 2;
}

message UrgenceRequest {
  TypeUrgence type = 1;
  Coordonnees position = 2;
  string description = 3;
  string citoyen_id = 4; // Qui a lancé l'alerte
}

message UrgenceResponse {
  string id = 1; // ID MongoDB
  bool prise_en_compte = 2;
  string message = 3;
  StatutIntervention statut = 4;
}

message SuiviRequest {
  string id_urgence = 1;
}

message SuiviUpdate {
  Coordonnees position_actuelle = 1;
  string message_status = 2;
  int32 temps_estime_arrivee = 3; // Minutes
}

message HistoriqueRequest {
  string citoyen_id = 1; // Optionnel (filtre)
}

message HistoriqueResponse {
  repeated UrgenceRequest urgences = 1;
}